src/globus_sdk/experimental/session_error/session_error.py

```diff
@@ -0,0 +1,195 @@
+        session.
+    :vartype session_required_single_domain: list of str, optional
+
+    :ivar session_required_mfa: Whether MFA is required for the session.
+    :vartype session_required_mfa: bool, optional
+
+    :ivar session_required_scopes: A list of scopes for which consent is required.
+    :vartype session_required_scopes: list of str, optional
+
+    :ivar extra_fields: A dictionary of additional fields that were provided. May
+        be used for forward/backward compatibility.
+    :vartype extra_fields: dict
+    """
+
+    SUPPORTED_FIELDS = {
+        "session_message": str,
+        "session_required_identities": list,
+        "session_required_policies": list,
+        "session_required_single_domain": list,
+        "session_required_mfa": bool,
+        "session_required_scopes": list,
+    }
+
+    def __init__(
+        self,
+        session_message: t.Optional[str] = None,
+        session_required_identities: t.Optional[t.List[str]] = None,
+        session_required_policies: t.Optional[t.List[str]] = None,
+        session_required_single_domain: t.Optional[t.List[str]] = None,
+        session_required_mfa: t.Optional[bool] = None,
+        session_required_scopes: t.Optional[t.List[str]] = None,
+        **kwargs: t.Any,
+    ):
+        self.session_message: t.Optional[str] = session_message
+        self.session_required_identities = session_required_identities
+        self.session_required_policies = session_required_policies
+        self.session_required_single_domain: t.Optional[
+            t.List[str]
+        ] = session_required_single_domain
+        self.session_required_mfa: t.Optional[bool] = session_required_mfa
+        self.session_required_scopes: t.Optional[t.List[str]] = session_required_scopes
+        self.extra_fields: t.Dict[str, t.Any] = kwargs
+
+    @classmethod
+    def from_dict(
+        cls, param_dict: t.Dict[str, t.Any]
+    ) -> "GlobusSessionErrorAuthorizationParameters":
+        """
+        Instantiate from a session error authorization parameters dictionary. Raises
+        a ValueError if the dictionary does not contain a valid GlobusSessionError.
+
+        :param param_dict: The dictionary to create the error from.
+        :type param_dict: dict
+        """
+
+        # Enforce that the error_dict contains at least one of the fields we expect
+        if not any(field in param_dict for field in cls.SUPPORTED_FIELDS.keys()):
+            raise ValueError(
+                "Must include at least one supported authorization parameter: "
+                ", ".join(cls.SUPPORTED_FIELDS.keys())
+            )
+
+        # Enforce the field types
+        for field_name, field_type in cls.SUPPORTED_FIELDS.items():
+            if field_name in param_dict and not isinstance(
+                param_dict[field_name], field_type
+            ):
+                raise ValueError(f"{field_name} must be of type {field_type.__name__}")
```

Comment:
I actually implemented it in the constructor initially and then thought twice about it and changed my mind. I _think_ I could still be convinced otherwise, but there are potentially at least a couple issues.

Issue 1: If we just pass these to `__init__()` then calling `from_dict()` could raise either a `ValueError` or a `TypeError`, which means that we're potentially going to have to catch more types of errors in the utils methods, which means there's less precision and a greater chance we accidentally mask a programming error.

Issue 2: As a programmer manually instantiating this, you may want/need more flexibility, but there is higher trust here that you "know what you are doing" and it's not really the constructor's job to prevent you from instantiating at all (e.g., you might continue to modify these fields after instantiation and make it valid). If you are calling `from_dict` the assumption is that you're parsing something you've received from another service and trying to see if it's a proper session error, and in this case we act as strictly as possible.

I definitely feel less strongly about #2.

Reply:
> If you are calling `from_dict` the assumption is that you're parsing something you've received from another service

In that case, it seems like this logic should live in the session error utilities package. As is, `.from_dict()` and `to_session_error()` share similar intentions.

Reply:
> Issue 1: If we just pass these to __init__() then calling from_dict() could raise either a ValueError or a TypeError, which means that we're potentially going to have to catch more types of errors in the utils methods, which means there's less precision and a greater chance we accidentally mask a programming error.

Hmm, so it's about working around the fact that python treats a missing constructor argument as a TypeError (paired with the desire to have utils suppress errors on init). I hadn't thought about this line.

> Issue 2: As a programmer manually instantiating this, you may want/need more flexibility, but there is higher trust here that you "know what you are doing" and it's not really the constructor's job to prevent you from instantiating at all (e.g., you might continue to modify these fields after instantiation and make it valid). If you are calling from_dict the assumption is that you're parsing something you've received from another service and trying to see if it's a proper session error, and in this case we act as strictly as possible.

I'm less sympathetic to this line of thinking.
Reducing the verification done in construction reduces the guarantee of anyone interacting with the object.

If we want to allow users to build up an object which they finalize into a materialized object (which I'm not sure if we do), we should keep verification in the constructor but expose a builder pattern to explicitly defer verificiation until a certain point, not make our constructor a faux-builder while dropping it's verification:

```python
GlobusSessionError.builder()\
   .code("MyCoolError")\
   .session_authorization_parameter(parameter)\
   .build()
```
(this model/naming is ripped from the aws java sdk v2 [ref](https://github.com/awsdocs/aws-doc-sdk-examples/blob/main/javav2/example_code/dynamodb/src/main/java/com/example/dynamodb/GetItem.java#L58-L61))

---

Overall my two concerns are:
1. The one Stephen mentioned above - multiple paths to reason about lead to orphaned code (we update one path but not the other)
2. Callers not receiving the verification that they should because they used the standard constructor route.

What if we keep the key existence check in `from_dict(...)` for the sake of `TypeError` prevention but upstream any other checks into the constructor?